### PR TITLE
mixin-mixout: handle congestion corner case without assert

### PR DESCRIPTION
**Pull request by kv2019i**
If mixout is not able to process data, the "assert(sinks_free_frames >= start_frame)" condition is hit.

This is not correct usage of asserts as the condition may be hit due to performance reasons (and is frequently hit in pause-resume stress test with sof-test), and production builds will have asserts disabled, leading to undefined behaviour.

Instead of assert, add a check, emit a warning and truncate the amount of copied frames, to ensure deterministic behaviour in all conditions.

Link: https://github.com/thesofproject/sof/issues/6896
Signed-off-by: Kai Vehmanen <kai.vehmanen@linux.intel.com>

### CODE REVIEW

File: src/audio/mixin_mixout/mixin_mixout.c [hunk 1]
```diff
@@ -439,7 +439,11 @@ static int mixin_copy(struct comp_dev *dev)
 		 * performance problems with processing data on time in mixout.
 		 */
 		start_frame = mixout_data->pending_frames[source_index];
-		assert(sinks_free_frames >= start_frame);
+		if (start_frame > frames_to_copy) {
+			comp_warn(dev, "mixin copy: no space, need %u, space for %u",
+				  start_frame, frames_to_copy);
+			frames_to_copy = start_frame;
+		}
```
Comment (lyakh) [inline src/audio/mixin_mixout/mixin_mixout.c:446]: Do I understand it correctly, that previous the `assert()` was checking, that in the current sink the number of frames still pending from previous copies wasn't "too large," specifically that it wasn't larger than the number of free frames over all sinks? Which is similar to saying, that no sink is more than half full (assuming they all have the same size buffers)? Now you replace that check with a different one, namely checking that the number of still pending frames isn't larger than what we want to copy now? Which might actually be even easier to hit - if your last transfer is small and you still have a few pending frames from the previous run.

**Review by lgirdwood — COMMENTED** [review]
@kv2019i do we need to pass anything up the stack for host to take corrective action ?

File: src/audio/mixin_mixout/mixin_mixout.c (hunk 1 above)
Comment (kv2019i) [reply src/audio/mixin_mixout/mixin_mixout.c:446]: @lyakh  Correct. "frames_to_copy" already taken into account as "sinks_free_frames = MIN(sinks_free_frames, free_frames - pending_frames);" is done earlier in the function. 

So old check assumes "start_frame" cannot ever be larger than sinks_free_frames. This is not correct and you either have undefined behaviour (this currently happens) or assert is hit (when we enable CONFIG_ASSERT=y).

My modified code takes account both  how much we have leftovers in source, and how much we have space in output.

**Comment by kv2019i** [comment]
@lgirdwood wrote:
> @kv2019i do we need to pass anything up the stack for host to take corrective action ?

Not sure how to do this as the module really doesn't know why data flow is abnormal. The assert is mostly hit in pause-resume tests with lot of stream state transitions, so these are likely cases where the stream is anyways terminated. At least no failures are seen with this patch. At least this is better than running the process() with invalid parameters (start_frames > space in sinks).

@ranj063 is working on the https://github.com/thesofproject/sof/pull/6836 , but does not seem to touch this part specifically.

**Review by lgirdwood — APPROVED** [review]

**Comment by lgirdwood** [comment]
@lyakh good for you ?
@kv2019i do we need any other PRs to be merged before this will pass CI or is the failure unexpected ?

**Comment by kv2019i** [comment]
@lgirdwood I need to spend more time and respond to @lyakh 's concern. Let's not merge before that.

**Comment by gkbldcig** [comment]
Can one of the admins verify this patch?

**Comment by kv2019i** [comment]
V2 (Feb 3rd):
- modified patch based on review feedback
- it turns out the assert condition is in fact incorrect and can simply be removed

File: src/audio/mixin_mixout/mixin_mixout.c (hunk 1 above)
Comment (kv2019i) [reply src/audio/mixin_mixout/mixin_mixout.c]: @lyakh Boy, this took a long time. But thanks, you are right, my fix was not correct. Looking at this again, it seems the problem is with the assert and it's actually not needed. If "start_frame == 10" and "frames_to_copy == 5", the "sinks_free_frames" can still be "3", "5" or "10", or are valid values. The "start_frame" is specific to a single mixout buffer, while the other variables are calculcated as a minimum across all mixout buffers. So it's perfectly possible that a single mixout buffer has more already mixed data, than what can be written in this cycle by mixin. 

So long story short, the assert is not correct. And as the code can handle this case gracefully, no actions are needed in its place.

Comment (lyakh) [reply src/audio/mixin_mixout/mixin_mixout.c]: @kv2019i mmh, almost. I think `sinks_free_frames == 3` would be invalid in your above example, but that's ok because it cannot be that. IIUC in mixin copying method we look over all mixouts, connected to this mixin, check free space in each of them for us and select the minimum over them all so that we copy the same number of frames to all to make sure that the mixin itself has a consistent state - it has pushed out `frames_to_copy` frames to all mixouts and its number of available for copying frames is reduced by that number. But otherwise yes, that assertion does seem wrong to me too. Maybe it got here by confusion with https://github.com/thesofproject/sof/blob/a88ba51bb4c5cd6b7a984179d39d1b9110e70d94/src/audio/mixin_mixout/mixin_mixout.c#L368 where a similarly names `free_frames` variable means something different - it should include pending frames and therefore cannot be smaller than those.

Comment (kv2019i) [reply src/audio/mixin_mixout/mixin_mixout.c]: Right @lyakh , so what I mean that if "sinks_free_frames == 3", the existing code witll truncate frames_to_copy so it will be limited to 3 as well (even if "source_avail_frames == 5").

So we are good to go?

Comment (lyakh) [reply src/audio/mixin_mixout/mixin_mixout.c]: @kv2019i well, yeah, we can commit this, but there's a bigger problem with mixin-mixout that I dug out while working on #6864 : https://github.com/thesofproject/sof/issues/6864#issuecomment-1419330298 Any ideas?

**Review by lyakh — APPROVED** [review]

File: src/audio/mixin_mixout/mixin_mixout.c (hunk 1 above)
Comment (kv2019i) [reply src/audio/mixin_mixout/mixin_mixout.c]: Oh that #6864 is interesting. But let's continue over there. I'd rather get this merged soon so we can enable assert and gets the "gates up" in our CI.

**Comment by kv2019i** [comment]
V3 (Feb 7th):
- no functional changes, just plain rebase

**Review by lgirdwood — APPROVED** [review]

**Review by dnikodem — APPROVED** [review]
